docs/branching.md

```diff
@@ -0,0 +1,98 @@
+# Branching and deployments model
+
+This document aims to discuss and document the relations between branches and deployments of `elixir-omg`, with respect to branches and deployments of `plasma-contracts`.
+
+It is a refinement of the [OIP4 branching model](https://github.com/omisego/OIP/blob/master/0004-ewallet-release-and-versioning.md), applicable to `elixir-omg` and `plasma-contracts` versioning.
+
+Rationale:
+- the history and the relations between the versions are readable and simple to understand
+- we can predictably sync our respective watchers/run child chain servers against deployed contracts
+- we can move with the `master` branch quickly
+
+
+## Dependency Rules for elixir-omg
+
+### For mix.exs
+- `elixir-omg/master` will point to the `plasma-contracts/master`
+- A release branch in `elixir-omg` will point to the corresponding release branch in `plasma-contracts`. For example:
+  - `elixir-omg/v0.1` -> `plasma-contracts/v0.1`
+
+### For mix.lock
+- Always points to a specific SHA that in the history of the `plasma-contracts` branch referenced in `mix.exs`
+
+
+## Deployment Scenarios
+
+### 1 - Single production deployment, ongoing development
+
+This is the active scenario most of the time.
+
+Branches and environments:
+- `master` is automatically deployed to **development** environment
+- `v0.1` is automatically deployed to **staging-v0-1** environment
+  - changes to the release branch will be merged into `master`, as needed
+- `v0.1` is manually deployed to **production-v0-1** environment
+
+Deploying new contracts in `master`:
+- make a PR to `elixir-omg/master` bumping the contract version in `mix.lock`
+- CI checks on the new integration
+- merge the PR
+- redeploy contracts
+- redeploy child chain and watcher
+- NB – contract deployment is currently a manual process, so we may be in a state where `mix.lock` points to a newer SHA than deployed on **development**. _We will correct this disparity as quickly as possible. This may mean rolling back `mix.lock`, if needed._
+- TODO – Automate contract deployments in **development**
+
+Deploying new contracts in the release branch:
+- :stop_sign: - _NOPE_
+- We cannot deploy any `elixir-omg` code that is incompatible with the currently deployed contracts in **staging** and **production**
+
+### 2 - Production deployment, validating a new version for network upgrade
+
+Branches and environments:
+- `master` is automatically deployed to **development** environment
```

Comment:
If we're deploying the same code to two different environments couldn't we consolidate `development` and `staging` into once cluster? That would save on infrastructure costs. We can still have the concept of the two but they'd live on one cluster.

Reply:
I don't think this is the same code. `staging` get's deployed from `v0.2` pushes, so significantly less frequently and, starting from cutting point, from different code entirely.

Reply:
As we're on `v0.2` now, it might be good to include that in this document too.

Reply:
in this context `v0.1` and `v0.2` are used as "two example stable branches". The original doc from #692 had:

> This takes `v0.1` and `v0.2` as an example of two versions, where one is deployed in `staging`/`Ari` and the other is the upcoming upgrade and is deployed in `development`.

maybe we should add such (possibly groomed) explanation at the top?

And to answer your comment specifically - in this "phase" - "Production deployment, validating a new version for network upgrade", there is not yet a `v0.2` branch cut, the development towards `v0.2` is done in `master` branch

Reply:
The thought here is to deploy the same code (`master`) to two environments. The assumption was that **staging** is as close to **production** as possible (resources, tooling, monitoring, etc.), so we get a more realistic sense of how `master` is going to perform. This is more of an infrastructure validation step. @jbunce – certainly open to an argument about this not being necessary.

The other goal here is to delay cutting the new release branch (in this example, `v0.2`) as long as possible, so that we minimize the period we may need to port changes between `master` and `v0.2`. As stated in "Scenario 3", we cut the release branch when we're ready to provision the new **production** environment, which looks like how things went with Samrong 🙌.

_NB – Acknowledged that as the team grows and we work on features more concurrently, we'll probably have to cut the release branch sooner, which puts us in perhaps a "Scenario 2.5" – But I I'm trying to keep it simple here._

Reply:
Why deploy the same code to two different places? For me this:

- Adds more complexity in remembering which environment contract you deposited into
- Creates redundant deployments that aren't used, therefore burning unnecessary cycles (and therefore money)

I can definitely see a use case once the `v0.2` branch is cut to have:

`master` -> `development`
`v0.2` -> staging

Perhaps until then we can roll with a single deployment environment in `development`?

Reply:
> Why deploy the same code to two different places?

Assuming that *staging* mirrors *production* more closely than *development*, the purpose is to vet the code against a *production*-like infrastructure. It's certainly an intermediary step.

But perhaps I'm being too cautious and overly concerned with managing two branches. If it makes things simpler, I'm fine to deploy *staging* once we cut a new release branch. I'll update to reflect that.